tls/s2n_cipher_suites.c

```diff
@@ -1155,24 +1155,138 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+        *chacha20_boosted = true;
+        return S2N_RESULT_OK;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/* Checks that a mutually-supported cipher suite can be used in this connection */
+static S2N_RESULT s2n_validate_cipher_suite_match(struct s2n_connection* conn, struct s2n_cipher_suite** potential_match, bool* match_is_valid) {
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(match_is_valid);
+    RESULT_ENSURE_REF(potential_match);
+
+    *match_is_valid = false;
+
+    /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+    if ((conn->actual_protocol_version >= S2N_TLS13) != ((*potential_match)->minimum_required_tls_version >= S2N_TLS13)) {
+        return S2N_RESULT_OK;
+    }
+
+    /* If connection is for SSLv3, use SSLv3 version of suites */
+    if (conn->actual_protocol_version == S2N_SSLv3) {
+        *potential_match = (*potential_match)->sslv3_cipher_suite;
+    }
+
+    /* Skip the suite if we don't have an available implementation */
+    if (!(*potential_match)->available) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Make sure the cipher is valid for available certs */
+    if (s2n_is_cipher_suite_valid_for_auth(conn, *potential_match) != S2N_SUCCESS) {
+        return S2N_RESULT_OK;
+    }
+
+    /* TLS 1.3 does not include key exchange in cipher suites */
+    if ((*potential_match)->minimum_required_tls_version < S2N_TLS13) {
+        /* If the kex is not supported continue to the next candidate */
+        bool kex_supported = false;
+        if (s2n_result_is_error(s2n_kex_supported(*potential_match, conn, &kex_supported))) {
+            return S2N_RESULT_OK;
+        }
+
+        if (!kex_supported) {
+            return S2N_RESULT_OK;
+        }
+
+        /* If the kex is not configured correctly continue to the next candidate */
+        if (s2n_result_is_error(s2n_configure_kex(*potential_match, conn))) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    /**
+    *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+    *# The server MUST ensure that it selects a compatible PSK
+    *# (if any) and cipher suite.
+    **/
+    if (conn->psk_params.chosen_psk != NULL) {
+        if ((*potential_match)->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    *match_is_valid = true;
+    return S2N_RESULT_OK;
+}
+
 static int s2n_set_cipher_as_server(struct s2n_connection *conn, uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
 {
     POSIX_ENSURE_REF(conn);
     POSIX_ENSURE_REF(conn->secure);
 
+    struct s2n_blob wire_blob = {0};
+    POSIX_GUARD(s2n_blob_init(&wire_blob, wire + cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN, count * cipher_suite_len));
+
+    struct s2n_stuffer wire_stuffer = {0};
+    POSIX_GUARD(s2n_stuffer_init(&wire_stuffer, &wire_blob));
+    POSIX_GUARD(s2n_stuffer_write(&wire_stuffer, &wire_blob));
```

Comment:
`wire_stuffer` doesn't allocate any memory, so you don't actually need to call s2n_stuffer_free. If you did, you'd want to use DEFER_CLEANUP instead of trying to remember to call it on every possible return branch.